File: bst/lb-10-convert-normal-bst-balanced-bst.py
# Balanced bst is the bst where height difference is atmost 1.
def find_inorder(root):
    arr = []

    def find(root):
        if not root:
            return
        find(root.left)
        arr.append(root.data)
        find(root.right)
    find(root)
    return arr


def convert_balanced_bst(root):
    # find the inorder traversal of tree.
    # divide in the middle of inorder traversals.
    # now create left, right child form sub tree.
    ino = find_inorder(root)
    # already bst
    # ino = sorted(ino) 

    def helper(arr, start, end):
        if start > end:
            return None
        mid = start + int((end-start)//2)
        root = Node(arr[mid])
        root.left = helper(arr, start, mid-1)
        root.right = helper(arr, mid+1, end)
        return root
    return helper(root, 0, len(ino)-1)

File: bst/cpp/lb-10-convert-normal-bst-balanced-bst.cpp
void find_inorder(Node *root, vector<int> &ino) {
    if (!root) return;
    find_inorder(root->left, ino);
    ino.push_back(root->data);
    find_inorder(root->right, ino);
}
Node *helper(vector<int> &arr, int start, int end) {
    if (start > end) return NULL;
    int mid = start + int((end - start) / 2);
    Node *root = new Node(arr[mid]);
    root->left = helper(arr, start, mid - 1);
    root->right = helper(arr, mid + 1, end);
    return root;
}
Node *convert_balanced_bst(Node *root) {
    // # find the inorder traversal of tree.
    // # divide in the middle of inorder traversals.
    // # now create left, right child form sub tree.
    vector<int> arr;
    find_inorder(root, arr);
    // sorting is not required as it is already a sbst
    // sort(arr.begin(), arr.end());

    return helper(root, 0, arr.size() - 1);
}
